### PR TITLE
Hanfei/rewrite plan

### PR DESCRIPTION
**Pull request by hanfei19910905**
remove result field from plan stage.

### CODE REVIEW

**Comment by hanfei19910905** [comment]
@shenli @coocood @zimulala  PTAL

File: evaluator/builtin_other.go [hunk 1]
```diff
@@ -0,0 +1,327 @@
+package evaluator
```
Comment (shenli) [inline evaluator/builtin_other.go]: Miss license.

File: executor/builder.go [hunk 2]
```diff
@@ -112,14 +113,14 @@ func (b *executorBuilder) build(p plan.Plan) Executor {
...
-func composeCondition(conditions []ast.ExprNode) ast.ExprNode {
+func composeCondition(conditions []expression.Expression) expression.Expression {
 	length := len(conditions)
 	if length == 0 {
 		return nil
 	} else if length == 1 {
 		return conditions[0]
 	} else {
-		return &ast.BinaryOperationExpr{Op: opcode.AndAnd, L: composeCondition(conditions[:length/2]), R: composeCondition(conditions[length/2:])}
+		return expression.NewFunction(model.NewCIStr("eq"), []expression.Expression{composeCondition(conditions[0 : length/2]), composeCondition(conditions[length/2:])})
```
Comment (coocood) [inline executor/builder.go]: Use constants to define function name like 'eq`.

File: expression/expression.go [hunk 3]
```diff
@@ -0,0 +1,366 @@
...
+	aggrMap  map[*ast.AggregateFuncExpr]int
+}
+
+// Expression represents all scalar expression in SQL.
+type Expression interface {
+	// Eval evaluates a expression through a row.
+	Eval(row []types.Datum, ctx context.Context) (types.Datum, error)
+
+	// Get the expression return type.
+	GetType() types.FieldType
```
Comment (shenli) [inline expression/expression.go]: Returning *types.FieldType will save memory allocation.

File: expression/expression.go [hunk 4]
```diff
@@ -0,0 +1,366 @@
...
+// DeepCopy implements Expression interface.
+func (col *Column) DeepCopy() Expression {
+	newCol := *col
+	return &newCol
+}
+
+// Schema stands for the row schema get from input.
+type Schema []*Column
+
+// FindColumn replace a ast column to a expression column.
```
Comment (shenli) [inline expression/expression.go]: replaces an ast column with an expression column

File: optimizer/plan/column_pruning.go [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+func retrieveIndex(cols []*expression.Column, schema expression.Schema) {
+	for _, col := range cols {
+		idx := schema.GetIndex(col)
+		if col.Index == -1 {
+			col.Index = idx
+		}
+	}
+}
+
+// ColumnPruning prune unused column and find index for columns.
```
Comment (shenli) [inline optimizer/plan/column_pruning.go]: prune -> prunes
find -> finds

File: optimizer/plan/new_plans.go [hunk 6]
```diff
@@ -39,10 +40,62 @@ type Join struct {
...
+	OtherConditions []expression.Expression
+}
+
+// Projection represents a select fields plan.
+type Projection struct {
+	basePlan
+	exprs []expression.Expression
+}
+
+// Aggregation represents a aggregate plan.
```
Comment (shenli) [inline optimizer/plan/new_plans.go]: a->an

File: optimizer/plan/newplanbuilder.go [hunk 7]
```diff
@@ -130,32 +177,274 @@ func (b *planBuilder) buildNewJoin(join *ast.Join) Plan {
...
+	orderByAggrFuncs := extractor.AggFuncs
+	extractor.AggFuncs = make([]*ast.AggregateFuncExpr, 0)
+	orderByMapper := make(map[*ast.AggregateFuncExpr]int)
+	for _, aggr := range orderByAggrFuncs {
+		orderByMapper[aggr] = len(sel.Fields.Fields)
+		field := &ast.SelectField{Expr: aggr, AsName: model.NewCIStr(fmt.Sprintf("sel_aggr_%d", len(sel.Fields.Fields)))}
+		sel.Fields.Fields = append(sel.Fields.Fields, field)
+	}
+
+	for _, f := range sel.GetResultFields() {
```
Comment (coocood) [inline optimizer/plan/newplanbuilder.go]: FieldList

File: expression/expression.go [hunk 8]
```diff
@@ -0,0 +1,366 @@
...
+type expressionRewriter struct {
+	ctxStack []Expression
+	schema   Schema
+	err      error
+	aggrMap  map[*ast.AggregateFuncExpr]int
+}
+
+// Expression represents all scalar expression in SQL.
+type Expression interface {
+	// Eval evaluates a expression through a row.
```
Comment (shenli) [inline expression/expression.go]: an expression

File: optimizer/plan/newplanbuilder.go [hunk 9]
```diff
@@ -106,18 +133,38 @@ func extractOnCondition(conditions []ast.ExprNode, left Plan, right Plan) (eqCon
 			otherCond = append(otherCond, expr)
 		}
 	}
-	return eqCond, leftCond, rightCond, otherCond
+	return
+}
+
```
Comment (coocood) [inline optimizer/plan/newplanbuilder.go]: Add comments about this function.

File: optimizer/plan/column_pruning.go [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+
+package plan
+
+import (
+	"fmt"
+	"github.com/juju/errors"
+	"github.com/pingcap/tidb/expression"
+)
+
+func retrieveIndex(cols []*expression.Column, schema expression.Schema) {
```
Comment (coocood) [inline optimizer/plan/column_pruning.go]: rename to resolveIndex would be better.

File: optimizer/plan/column_pruning.go [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+		idx := schema.GetIndex(col)
+		if col.Index == -1 {
+			col.Index = idx
+		}
+	}
+}
+
+// PruneColumnsAndResolveIndices prunes unused columns and resolves index for columns.
+func PruneColumnsAndResolveIndices(p Plan) error {
+	//TODO: Currently we only implement index retrieving, column pruning will be implemented later.
```
Comment (coocood) [inline optimizer/plan/column_pruning.go]: index resolving.

**Comment by coocood** [comment]
Rest LGTM

**Comment by coocood** [comment]
@shenli PTAL

File: expression/expression.go [hunk 12]
```diff
@@ -119,7 +119,7 @@ func (col *Column) DeepCopy() Expression {
 // Schema stands for the row schema get from input.
 type Schema []*Column
 
-// FindColumn replace a ast column to a expression column.
+// FindColumn replaces an ast column to an expression column.
```
Comment (shenli) [inline expression/expression.go]: replace sth with sth

File: expression/expression.go [hunk 13]
```diff
@@ -305,48 +308,23 @@ func (er *expressionRewriter) Leave(inNode ast.Node) (retNode ast.Node, ok bool)
...
 			er.err = errors.New("Can't find function!")
 			return retNode, false
 		}
 		function.function = f.F
-		function.retType.Tp = types.KindInt64
 		er.ctxStack = er.ctxStack[:length-2]
 		er.ctxStack = append(er.ctxStack, function)
 	case *ast.UnaryOperationExpr:
-		function := &ScalarFunction{Args: []Expression{er.ctxStack[length-1]}}
+		function := &ScalarFunction{Args: []Expression{er.ctxStack[length-1]}, retType: v.Type}
```
Comment (shenli) [inline expression/expression.go]: Forget to set function.function?

File: expression/expression.go [hunk 14]
```diff
@@ -0,0 +1,350 @@
...
+			idx = i
+		}
+	}
+	if idx == -1 {
+		return nil, fmt.Errorf("Unknown column %s %s %s.", dbName.L, tblName.L, colName.L)
+	}
+	return s[idx], nil
+}
+
+// GetIndex finds the index for a column
```
Comment (shenli) [inline expression/expression.go]: Miss '.' at the end.

File: expression/expression.go [hunk 15]
```diff
@@ -0,0 +1,350 @@
...
+
+// Eval implements Expression interface.
+func (sf *ScalarFunction) Eval(row []types.Datum, ctx context.Context) (types.Datum, error) {
+	args := make([]types.Datum, len(sf.Args))
+	for _, arg := range sf.Args {
+		result, err := arg.Eval(row, ctx)
+		if err != nil {
+			args = append(args, result)
+		} else {
+			return types.Datum{}, err
```
Comment (shenli) [inline expression/expression.go]: return errors.Trace(err)

File: expression/expression.go [hunk 16]
```diff
@@ -0,0 +1,350 @@
...
+			return retNode, false
+		}
+		function.function = f.F
+		er.ctxStack = er.ctxStack[:length-1]
+		er.ctxStack = append(er.ctxStack, function)
+	case *ast.BinaryOperationExpr:
+		function := &ScalarFunction{Args: []Expression{er.ctxStack[length-2], er.ctxStack[length-1]}, retType: v.Type}
+		funcName, ok := opcode.Ops[v.Op]
+		if !ok {
+			er.err = fmt.Errorf("Unknown opcode %v", v.Op)
```
Comment (shenli) [inline expression/expression.go]: Use errors.New()

File: expression/expression.go [hunk 17]
```diff
@@ -0,0 +1,350 @@
...
+		f, ok := evaluator.Funcs[function.FuncName.L]
+		if !ok {
+			er.err = errors.New("Can't find function!")
+			return retNode, false
+		}
+		function.function = f.F
+		er.ctxStack = er.ctxStack[:length-1]
+		er.ctxStack = append(er.ctxStack, function)
+	default:
+		er.err = fmt.Errorf("UnkownType: %T", v)
```
Comment (shenli) [inline expression/expression.go]: Use errors.New()

File: optimizer/plan/column_pruning.go [hunk 18]
```diff
@@ -0,0 +1,85 @@
...
+			cols = extractColumn(eqCond, cols)
+		}
+		for _, eqCond := range v.RightConditions {
+			cols = extractColumn(eqCond, cols)
+		}
+		for _, eqCond := range v.OtherConditions {
+			cols = extractColumn(eqCond, cols)
+		}
+	default:
+		return fmt.Errorf("Unknown Type %T", v)
```
Comment (shenli) [inline optimizer/plan/column_pruning.go]: errors.New

File: optimizer/plan/column_pruning.go [hunk 19]
```diff
@@ -0,0 +1,85 @@
...
+	}
+	for _, child := range p.GetChildren() {
+		retrieveIndex(cols, child.GetSchema())
+	}
+	for _, col := range cols {
+		if col.Index == -1 {
+			return fmt.Errorf("Can't find column %s", col.ColName.L)
```
Comment (shenli) [inline optimizer/plan/column_pruning.go]: errors.New()

File: optimizer/plan/predicate_push_down.go [hunk 20]
```diff
@@ -170,18 +151,25 @@ func PredicatePushDown(p Plan, predicates []ast.ExprNode) (ret []ast.ExprNode, e
...
 			}
+			if len(retCond) != 0 {
+				addFilter(v, proj, retCond)
+			}
 		}
-		return predicates, nil
+		return
+	//TODO: support aggregation.
+	default:
+		return predicates, fmt.Errorf("Unkown type %T.", v)
```
Comment (shenli) [inline optimizer/plan/predicate_push_down.go]: errors.New()

File: expression/expression.go [hunk 21]
```diff
@@ -0,0 +1,350 @@
...
+
+// Rewrite rewrites ast to Expression.
+func Rewrite(expr ast.ExprNode, schema Schema, AggrMapper map[*ast.AggregateFuncExpr]int) (newExpr Expression, err error) {
+	er := &expressionRewriter{schema: schema, aggrMap: AggrMapper}
+	expr.Accept(er)
+	if er.err != nil {
+		return nil, errors.Trace(er.err)
+	}
+	if len(er.ctxStack) != 1 {
+		return nil, fmt.Errorf("context len %v is invalid", len(er.ctxStack))
```
Comment (shenli) [inline expression/expression.go]: errors.New

File: expression/expression.go [hunk 22]
```diff
@@ -0,0 +1,350 @@
...
+type Schema []*Column
+
+// FindColumn replaces an ast column with an expression column.
+func (s Schema) FindColumn(astCol *ast.ColumnName) (*Column, error) {
+	dbName, tblName, colName := astCol.Schema, astCol.Table, astCol.Name
+	idx := -1
+	for i, col := range s {
+		if (dbName.L == "" || dbName.L == col.DbName.L) && (tblName.L == "" || tblName.L == col.TblName.L) && (colName.L == col.ColName.L) {
+			if idx != -1 {
+				return nil, fmt.Errorf("Column '%s' is ambiguous", colName.L)
```
Comment (shenli) [inline expression/expression.go]: errors.New

File: expression/expression.go [hunk 23]
```diff
@@ -0,0 +1,350 @@
...
+			}
+			idx = i
+		}
+	}
+	if idx == -1 {
+		return nil, fmt.Errorf("Unknown column %s %s %s.", dbName.L, tblName.L, colName.L)
```
Comment (shenli) [inline expression/expression.go]: errors.New

File: optimizer/plan/stringer.go [hunk 24]
```diff
@@ -94,6 +94,21 @@ func toString(in Plan, strs []string, idxs []int) ([]string, []int) {
...
+		strs = strs[:idx]
+		str = "UnionAll{" + strings.Join(children, "->") + "}"
+		idxs = idxs[:last]
+	case *NewTableScan:
+		str = fmt.Sprintf("DataScan(%v)", x.Table.Name.L)
+	case *Selection:
+		str = "Selection"
+	case *Projection:
+		str = "Projection"
+	case *Aggregation:
```
Comment (shenli) [inline optimizer/plan/stringer.go]: What's the different between Aggregation and Aggregate?
See line 112 and 110.

**Comment by shenli** [comment]
LGTM

**Pull request merged**